onnxruntime/contrib_ops/cpu/reverse.cc

```diff
@@ -0,0 +1,300 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "reverse.h"
+#include <utility>
+#include "core/util/math.h"
+#include "core/util/math_cpuonly.h"
+#include "core/framework/utils.h"
+
+namespace onnxruntime {
+namespace contrib {
+
+ONNX_OPERATOR_KERNEL_EX(
+    Reverse,
+    kMSDomain,
+    1,
+    kCpuExecutionProvider,
+    KernelDefBuilder().TypeConstraint("T", DataTypeImpl::AllTensorTypes()),
+    Reverse);
+
+// some templated aliases
+template <typename T, int rank>
+using EigenTensorMap = Eigen::TensorMap<Eigen::Tensor<T, rank, Eigen::RowMajor, Eigen::DenseIndex>>;
+
+template <typename T, int rank>
+using ConstEigenTensorMap = Eigen::TensorMap<Eigen::Tensor<const T, rank, Eigen::RowMajor, Eigen::DenseIndex>>;
+
+// utility helpers specific to Reverse
+template <int rank>
+Eigen::array<bool, rank> generate_bool_reverse_axes(const std::vector<int64_t>& reverse_axes) {
+  Eigen::array<bool, rank> eigen_reverse_axes;
+
+  // default axes - reverse all axes as per spec
+  if (reverse_axes.size() == 0) {
+    for (int i = 0; i < rank; ++i) {
+      eigen_reverse_axes[i] = true;
+    }
+    return eigen_reverse_axes;
+  }
+
+  // explicit axes given
+  eigen_reverse_axes.fill(false);
+  for (int i = 0; i < reverse_axes.size(); ++i) {
+    const auto& dim = reverse_axes[i];
+    eigen_reverse_axes[dim >= 0 ? dim : dim + rank] = true;
+  }
+  return eigen_reverse_axes;
+}
+
+template <int rank>
```

Comment:
><int rank> [](start = 9, length = 10)

Why duplicate the code on the integer instead of passing it as an arg?

Reply:
Ditto everywhere in this code. 

---
In reply to: [269219970](https://github.com/Microsoft/onnxruntime/pull/703#discussion_r269219970) [](ancestors = 269219970)

Reply:
We leverage Eigen class instances that need to know the type, rank, and shape at compile-time. See here - https://eigen.tuxfamily.org/dox-devel/unsupported/eigen_tensors.html. How can we do this by passing in rank dynamically ?